src/cls/IPM/ResourceProcessor/WebApplication.cls

```diff
@@ -0,0 +1,263 @@
+        } ElseIf ((pPhase = "Unconfigure") || ((pPhase = "Clean") && tIsApplication)) {
+            // Remove web application
+            Do ..DeleteWebApp(tVerbose)
+        }
+    } Catch e {
+        Set sc = e.AsStatus()
+    }
+    Quit sc
+}
+
+Method CreateOrUpdateWebApp(pVerbose As %Boolean = 0) [ Internal ]
+{
+    // TODO: Make mirror-safe.
+    Set tOriginalNS = $namespace
+    Set dbDir = $$$defdir
+    New $namespace
+    Set $namespace = "%SYS"
+    
+    Merge properties = ..SecurityProperties
+    If $D(properties("MatchRoles"), tMatchRoles) # 2 {
+        Set properties("MatchRoles") = ..ReplaceMatchRoles(tMatchRoles, dbDir)
+    }
+    If $D(properties("NameSpace"), tNameSpace) {
+        Set properties("NameSpace") = ..ReplaceNameSpace(tNameSpace, tOriginalNS)
+    } Else {
+        // Does WSGI application need a namespace?
+        Set properties("NameSpace") = tOriginalNS
+    }
+
+    If ##class(Security.Applications).Exists(..Name) {
+        Write:pVerbose !,"Updating Web Application ",..Name
+        Set sc = ##class(Security.Applications).Get(..Name,.oldProperties)
+        $$$ThrowOnError(sc)
+        
+        Kill changes
+        Set key = ""
+        For {
+            Set key = $Order(properties(key),1,value)
+            Quit:(key="")
+            Set oldValue = $Get(oldProperties(key))
+            If (value '= oldValue) {
+                If (value = "") {
+                    Set value = "[missing]"
+                }
+                If (oldValue = "") {
+                    Set oldValue = "[missing]"
+                }
+                Set changes($i(changes)) = key_": "_oldValue_" -> "_value 
+            }
+        }
+        
+        If $Data(changes) {
+            If (pVerbose) {
+                For i=1:1:$Get(changes) {
+                    Write !,changes(i)
+                }
+            }
+            Set sc = ##class(Security.Applications).Modify(..Name,.properties)
+            $$$ThrowOnError(sc)
+            Write:pVerbose !,"Done."
+        } Else {
+            Write:pVerbose !,"No changes detected or made."
+        }
+    } Else {
+        Write:pVerbose !,"Creating Web Application ",..Name
+        Set sc = ##class(Security.Applications).Create(..Name,.properties)
+        If (pVerbose) {
+            Set key = ""
+            For {
+                Set key = $Order(properties(key),1,tValue)
+                Quit:key=""
+                Write !,key,": ",tValue
+            }
+        }
+        $$$ThrowOnError(sc)
+        Write:pVerbose !,"Done."
+    }
+}
+
+/// This removes an existing CSP application
+Method DeleteWebApp(pVerbose As %Boolean = 0) [ Internal ]
+{
+    // TODO: Make mirror-safe.
+    
+    // Only try to purge files if the directory exists.
+    Set sc = ##class(%Library.EnsembleMgr).deletePortal(..Name,"",pVerbose)
+    Set errorCodes = $System.Status.GetErrorCodes(sc)
+    If (errorCodes [ $$$ApplicationDoesNotExist || errorCodes [ $$$DeleteObjectNotFound) {
```

Comment:
ObjectScript operator precedence: this doesn't do what you think it does. Change to:
`If ((errorCodes [ $$$ApplicationDoesNotExist) || (errorCodes [ $$$DeleteObjectNotFound))`

Reply:
Fixed.